radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java

```diff
@@ -0,0 +1,225 @@
+    this.currentSubStateAccumulatorHash =
+        getPreviousSubStateAccumulatorHash(null).orElse(new byte[0]);
+    this.epoch = getEpoch();
+  }
+
+  @Override
+  public void close() {
+    this.subStateAccumulatorHashDatabase.close();
+    this.epochHashDatabase.close();
+  }
+
+  @Override
+  public void process(
+      Transaction dbTxn,
+      REProcessedTxn txn,
+      long stateVersion,
+      Function<SystemMapKey, Optional<RawSubstateBytes>> mapper) {
+
+    watch.start();
+    var isEpochChange = false;
+    var subStateBytes = new byte[0];
+    for (REStateUpdate reStateUpdate : txn.stateUpdates().toList()) {
+      subStateBytes = Arrays.concatenate(subStateBytes, getBytes(reStateUpdate));
+      if (reStateUpdate.getParsed() instanceof EpochData epochData) {
+        epoch = epochData.epoch();
+        isEpochChange = true;
+      }
+    }
+
+    this.currentSubStateAccumulatorHash =
+        HashUtils.sha256(Arrays.concatenate(this.currentSubStateAccumulatorHash, subStateBytes))
+            .asBytes();
+    var key = new DatabaseEntry(SUBSTATE_ACCUMULATOR_HASH);
+    this.subStateAccumulatorHashDatabase.put(
+        dbTxn, key, new DatabaseEntry(this.currentSubStateAccumulatorHash));
+
+    if (isEpochChange) {
+      epochHashDatabase.put(
+          dbTxn,
+          new DatabaseEntry(Longs.toByteArray(epoch)),
+          new DatabaseEntry(this.currentSubStateAccumulatorHash));
+      if (logger.isInfoEnabled()) {
+        logger.info(
+            "Epoch Hash: {} for epoch {}. Time spent since last epoch: {} s.",
+            Bytes.toHexString(this.currentSubStateAccumulatorHash),
+            epoch,
+            watch.elapsed().toSeconds());
+      }
+      watch.reset();
+    }
+    if (watch.isRunning()) {
+      watch.stop();
+    }
+  }
+
+  private Optional<byte[]> getPreviousSubStateAccumulatorHash(Transaction dbTxn) {
+    var previousSubStateAccumulatorHash = new DatabaseEntry();
+    var key = new DatabaseEntry(SUBSTATE_ACCUMULATOR_HASH);
+    this.subStateAccumulatorHashDatabase.get(dbTxn, key, previousSubStateAccumulatorHash, null);
+    return Optional.ofNullable(previousSubStateAccumulatorHash.getData());
+  }
+
+  private long getEpoch() {
+    try (Cursor cursor = this.epochHashDatabase.openCursor(null, null)) {
+      var key = new DatabaseEntry();
+      var data = new DatabaseEntry();
+      cursor.getLast(key, data, null);
+      return key.getData() == null ? 0 : Longs.fromByteArray(key.getData());
+    }
+  }
+
+  private byte[] getBytes(REStateUpdate reStateUpdate) {
+    var op = reStateUpdate.isBootUp() ? new byte[] {0} : new byte[] {1};
+    var id = reStateUpdate.getId() != null ? reStateUpdate.getId().asBytes() : new byte[0];
+    var type = new byte[] {reStateUpdate.typeByte()};
+    var parsed = getParticleBytes(getParticle(reStateUpdate.getParsed()));
+    var stateBuf =
+        reStateUpdate.getStateBuf() != null
+            ? reStateUpdate.getRawSubstateBytes().getData()
```

Comment:
It seems that `getRawSubstateBytes` is a side-effecting operation: it consumes the bytes from the `stateBuf`. I wonder if there are any unobvious implications of this.

Reply:
hmh, or is this covered by the use of `Supplier` there :thinking:

Reply:
I think you are right, not sure Supplier makes any difference here. Actually, I don't know if we need to use this field, @talekhinezh ?

Reply:
@LukasGasior1 actually, it seems you are right about `Supplier` and a new `ByteBuffer` is returned every time. Please take a look at the new code.